### PR TITLE
fix(cfn-include): allow deploy-time values in Parameter substitutions in Fn::Sub expressions

### LINKED ISSUE #14047: (cloudformation-include): Incorrect rendering of Sub functions when using CfnInclude parameters
When importing a CloudFormation template using the CfnInclude class and specifying parameters with values that are the `valueAsString` property of instances of core CfnParameter, every instances of short sub functions (`!Sub`) are incorrectly translated. However, when the included template uses the long version of sub functions (`Fn::Sub`), the synthesized template is valid. 

### Reproduction Steps
**topic.yml**
```yaml
AWSTemplateFormatVersion: "2010-09-09"

Parameters:
  Environment:
    Type: String

Resources:
  SNSTopic:
    Type: AWS::SNS::Topic
    Properties: 
      TopicName: !Sub my-sns-topic-${Environment}

```
**short-sub-stack.ts**
```typescript
import * as cdk from '@aws-cdk/core';
import * as cfninc from '@aws-cdk/cloudformation-include'


export class ShortSubStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const environment = new cdk.CfnParameter(this, 'Environment', {
      type: 'String'
    })

    new cfninc.CfnInclude(this, 'UserPool', {
      templateFile: './lib/cfn-modules/topic.yml',
      parameters: {
        Environment: environment.valueAsString
      }
    })
  }
}
```
```bash
cdk synth --no-version-reporting --no-path-metadata --no-asset-metadata
```

### What did you expect to happen?

The resulting template's sub functions should either be a valid short sub, or a valid long sub as such: 
```yaml
Fn::Sub:
  - my-sns-topic-${Environment}
  - Environment:
      Ref: Environment
```

### What actually happened?

An invalid sub function is produced as such:

```yaml
Fn::Sub:
  Fn::Join:
    - ""
    - - my-sns-topic-
      - Ref: Environment
```

Resulting in the following cfn-lint error: `E1019: Sub should be a string or array of 2 items` 

And the following CloudFormation error when deployed: `An error occurred (ValidationError) when calling the CreateChangeSet operation: Template error: One or more Fn::Sub intrinsic functions don't specify expected arguments. Specify a string as first argument, and an optional second argument to specify a mapping of values to replace in the string`

### Environment

  - **CDK CLI Version  :1.97.0**
  - **Framework Version:**
  - **Node.js Version:v14.16.0**
  - **OS               : macOS Big Sur 11.2.3**
  - **Language (Version):TypeScript (4.2.3)**

### Other

When using a long sub function, this issue does not happen and the produced cloudformation is valid. The following `topic.yml` file can be used to reproduce this behaviour: 

```yaml
AWSTemplateFormatVersion: "2010-09-09"

Parameters:
  Environment:
    Type: String

Resources:
  SNSTopic:
    Type: AWS::SNS::Topic
    Properties: 
      TopicName: 
        Fn::Sub: 
          - my-sns-topic-${Environment}
          - Environment: !Ref Environment
```

--- 

This is :bug: Bug Report

### PR DESCRIPTION
**Pull request by skinny85**
When a Token that resolved to an object, like `{ Ref }`,
was passed as the value to be substituted for a CloudFormation Parameter,
and that Parameter was used in an Fn::Sub expression,
it caused incorrect CloudFormation to be generated
(an Fn::Join as an argument to Fn::Sub, which is not allowed).
Check for that case explicitly, and move that substitution into the Fn::Sub map instead.

Fixes #14047

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/cloudformation-include/test/valid-templates.test.ts [hunk 1]
```diff
@@ -326,6 +334,38 @@ describe('CDK Include', () => {
     );
   });
 
+  test('when a parameter in an Fn::Sub expression is substituted with a deploy-time value, it adds a new key to the Fn::Sub map', () => {
+    const parameter = new core.CfnParameter(stack, 'AnotherParam');
+    includeTestTemplate(stack, 'fn-sub-parameters.json', {
+      parameters: {
+        'MyParam': parameter.valueAsString,
```
Comment (rix0rrr) [inline packages/@aws-cdk/cloudformation-include/test/valid-templates.test.ts:341]: > it caused incorrect CloudFormation to be generated (an Fn::Join as an argument to Fn::Sub, which is not allowed).

I don't exactly know what this part of the PR description means, but it sounds foreboding.

Can we also test:

```
  parameters: {
    'MyParam': `its_a_${parameter.valueAsString}_concatenation`
  }
```

?

Does that do the right thing?

File: packages/@aws-cdk/core/lib/cfn-parse.ts [hunk 2]
```diff
@@ -682,48 +682,72 @@ export class CfnParser {
...
+      }
+
+      // lookup in map
+      if (refTarget in map) {
+        return leftHalf + '${' + refTarget + '}' + go(rightHalf);
+      }
+
+      // since it's not in the map, check if it's a pseudo-parameter
+      // (or a value to be substituted for a Parameter, provided by the customer)
+      const specialRef = self.specialCaseSubRefs(refTarget);
```
Comment (rix0rrr) [inline packages/@aws-cdk/core/lib/cfn-parse.ts:712]: Are pseudo-parameters even supported if there is an expression map?

I think as soon as there is a map, every key has to come from the map, no?

Comment (skinny85) [reply packages/@aws-cdk/core/lib/cfn-parse.ts:712]: No. There is no such limitation. This is valid CloudFormation:

```yaml
Resources:
  SNSTopic:
    Type: AWS::SNS::Topic
    Properties:
      TopicName:
        Fn::Sub:
          - my-sns-${Param}-${AWS::Region}-topic-${Environment}
          - Environment:
              Ref: AnotherParam
```

(where `Param` and `AnotherParam` are Parameters of the template)

File: packages/@aws-cdk/cloudformation-include/test/valid-templates.test.ts (hunk 1 above)
Comment (skinny85) [reply packages/@aws-cdk/cloudformation-include/test/valid-templates.test.ts]: Sure. Changed the test to do a concatenation.

**Comment by skinny85** [comment]
@rix0rrr included your comments, would appreciate another look!

**Review by rix0rrr — APPROVED** [review]

**Pull request merged**